gtsam/hybrid/HybridFactorGraph.cpp

```diff
@@ -0,0 +1,383 @@
+  // continuous variable, we make a GF if there are no hybrid factors;
+  // discrete variable, no continuous factor is allowed (escapes CG regime), so
+  // we panic, if discrete only we do the discrete elimination.
+
+  // However it is not that simple. During elimination it is possible that the
+  // multifrontal needs to eliminate an ordering that contains both Gaussian and
+  // hybrid variables, for example x1, c1. In this scenario, we will have a
+  // density P(x1, c1) that is a CLG P(x1|c1)P(c1) (see Murphy02)
+
+  // The issue here is that, how can we know which variable is discrete if we
+  // unify Values? Obviously we can tell using the factors, but is that fast?
+
+  // In the case of multifrontal, we will need to use a constrained ordering
+  // so that the discrete parts will be guaranteed to be eliminated last!
+
+  // Because of all these reasons, we need to think very carefully about how to
+  // implement the hybrid factors so that we do not get poor performance.
+  //
+  // The first thing is how to represent the GaussianMixture. A very possible
+  // scenario is that the incoming factors will have different levels of
+  // discrete keys. For example, imagine we are going to eliminate the fragment:
+  // $\phi(x1,c1,c2)$, $\phi(x1,c2,c3)$, which is perfectly valid. Now we will
+  // need to know how to retrieve the corresponding continuous densities for the
+  // assi- -gnment (c1,c2,c3) (OR (c2,c3,c1)! note there is NO defined order!).
+  // And we also need to consider when there is pruning. Two mixture factors
+  // could have different pruning patterns-one could have (c1=0,c2=1) pruned,
+  // and another could have (c2=0,c3=1) pruned, and this creates a big problem
+  // in how to identify the intersection of non-pruned branches.
+
+  // One possible approach is first building the collection of all discrete
+  // keys. After that we enumerate the space of all key combinations *lazily* so
+  // that the exploration branch terminates whenever an assignment yields NULL
+  // in any of the hybrid factors.
+
+  // When the number of assignments is large we may encounter stack overflows.
+  // However this is also the case with iSAM2, so no pressure :)
+
+  // PREPROCESS: Identify the nature of the current elimination
+  std::unordered_map<Key, DiscreteKey> discreteCardinalities;
+  std::set<DiscreteKey> discreteSeparatorSet;
+  std::set<DiscreteKey> discreteFrontals;
+
+  KeySet separatorKeys;
+  KeySet allContinuousKeys;
+  KeySet continuousFrontals;
+  KeySet continuousSeparator;
+
+  if (DEBUG) {
+    std::cout << RED_BOLD << "Begin Eliminate: " << RESET;
+    frontalKeys.print();
+  }
+
+  // This initializes separatorKeys and discreteCardinalities
+  for (auto &&factor : factors) {
+    if (DEBUG) {
+      std::cout << ">>> Adding factor: " << GREEN;
+      factor->print();
+      std::cout << RESET;
+    }
+    separatorKeys.insert(factor->begin(), factor->end());
+    if (!factor->isContinuous_) {
```

Comment:
Check for `isHybrid_`?

Reply:
Good idea, normally this would not happen so I'll add a debug guard also.

Reply:
Actually this accounts for both discrete and hybrid